core/src/main/scala/kafka/log/OffsetIndex.scala

```diff
@@ -114,13 +114,22 @@ class OffsetIndex(file: File, baseOffset: Long, maxIndexSize: Int = -1)
       OffsetPosition(relativeOffset(idx, n), physical(idx, n))
     }
   }
+
+  /**
+    * @param offset that we want to potentially append
+    * @return true if it is safe to append this offset to this index
+    */
+  def canAppend(offset: Long) : Boolean = {
```

Comment:
One more thing: having to call both `isFull` and `canAppend` methods in the `OffsetIndex` is not ideal. If we decide that a `canAppend` makes sense here, then it would probably be better for it to call `isFull` as well. In any case, let's hear Jun's reasoning for his suggestion first.